### PR TITLE
Add ROCm Clang and HIP to CUDA

### PR DESCRIPTION
**Pull request by cgmb**
This should now produce the sort of assembly that you'd expect to see when compiling HIP code. The release versions of the ROCm clang compilers should now be properly inlining all the GPU functions that you'd expect to be inlined. I'm not entirely sure how to make that happen for the vanilla clang trunk or for the ROCm clang trunk, so I left those two as they were for now.

The existing CUDA example should work, except in HIP you must add `#include <hip/hip_runtime.h>`.

This change depends on:
- https://github.com/compiler-explorer/misc-builder/pull/47

## TODO
- [x] Figure out how to fix these warnings:
```
error: Compiler ID clash for 'clang-rocm-trunk' - used by lang:c++ name:x86-64 clang (amd-stg-open), lang:cuda name:clang amd-stg-open (w/o device libs)
error: Compiler ID clash for 'clang-rocm-40502' - used by lang:c++ name:x86-64 clang rocm-4.5.2, lang:cuda name:clang rocm-4.5.2
error: Compiler ID clash for 'clang-rocm-50002' - used by lang:c++ name:x86-64 clang rocm-5.0.2, lang:cuda name:clang rocm-5.0.2
error: Compiler ID clash for 'clang-rocm-50103' - used by lang:c++ name:x86-64 clang rocm-5.1.3, lang:cuda name:clang rocm-5.1.3
error: Compiler ID clash for 'clang-rocm-50203' - used by lang:c++ name:x86-64 clang rocm-5.2.3, lang:cuda name:clang rocm-5.2.3
```
- [x] Figure out how to support multiple ISAs in the GUI rather than always passing `--offload-arch=gfx908`. There's _a lot_ of [AMDGPU ISAs](https://llvm.org/docs/AMDGPUUsage.html#processors) and one of the most useful things that CE could do is help people (like me) to understand how they differ from each other (and how they're the same).
- [x] Ensure the HIP library is available on the production image. For that purpose, I've opened https://github.com/compiler-explorer/infra/pull/834.

### CODE REVIEW

**Comment by cgmb** [comment]
@Artem-B did the initial work on adding HIP that I built on. I wonder if he has any thoughts.

Related issues:
- #2805
- #2794
- #2290

File: etc/config/cuda.amazon.properties
Comment (RubenRBS) [inline etc/config/cuda.amazon.properties:157]: Only bad thing about this is that compilers won't be ordered by semver using this prefix scheme. I have some ideas to fix it that will try to implement once this gets merged :)

**Review by RubenRBS — APPROVED** [review]

**Comment by cgmb** [comment]
I fixed the warnings:

    error: Compiler ID clash for 'clang-rocm-50203' - used by lang:c++ name:x86-64 clang rocm-5.2.3, lang:cuda name:clang rocm-5.2.3

by using the name `hipclang-rocm-50203` for the CUDA version of the compiler.

**Comment by cgmb** [comment]
I discovered that if you omit `--offload-arch=gfx908`, clang will build for for gfx803 by default (and it says the target architecture end of the assembly output, so users will be able to see that too).

This means that you don't need to do anything fancy to build for some other architecture. You just pass `--offload-arch=gfx908`, `--offload-arch=gfx90a`, `--offload-arch=gfx1030` or whatever you want.

**Comment by cgmb** [comment]
I added hip-amd to the libs section of the CUDA language config and have opened https://github.com/compiler-explorer/infra/pull/834 and https://github.com/compiler-explorer/misc-builder/pull/48 to do the rest of the setup.

**Comment by cgmb** [comment]
I dropped the nightly compiler builds from the GUI since they're not entirely functional. It seemed like it might be confusing to have a sort-of broken configuration listed as an option.

In any case, I believe this is now complete. It depends on the infra and misc-builder changes, but once those are in, I think this can be merged.

**Comment by mattgodbolt** [comment]
@RubenRBS any thoughts on the semver stuff or do you think we're OK to go ahead as-is?

**Comment by RubenRBS** [comment]
Sorry, missed this message. If everything else is good, lgtm and we can fix the semver once it's there

File: etc/config/cuda.amazon.properties
Comment (partouf) [inline etc/config/cuda.amazon.properties:154]: can you leave hiptrunk in here, but mark it as `compiler.hiptrunk.hidden=true`

Comment (cgmb) [reply etc/config/cuda.amazon.properties:154]: Sure. I've added hiptrunk back with all its original settings and just marked it as hidden.

**Review by mattgodbolt — APPROVED** [review]
The paths seem to check out! So I'm going to merge and push to our staging instance. If you're around right now I can have it up and running in ~15m. Else I'll kick the tyres and push to prod and you can see if it looks OK there. Let me know! :) We have a [Discord](https://discord.gg/B5WacA7) if that works too

**Comment by mattgodbolt** [comment]
Thanks so much for your patience on this! And thanks to @partouf for helping this along so much!

**Comment by cgmb** [comment]
> The paths seem to check out! So I'm going to merge and push to our staging instance. If you're around right now I can have it up and running in ~15m.

That's https://godbolt.org/staging ?

**Comment by mattgodbolt** [comment]
Yes! It will be anyway. I'll post here when it's up!

**Comment by mattgodbolt** [comment]
Staging is now up! Please let me know if it looks good to you: I can debug if not on my side!

**Comment by cgmb** [comment]
It appears to work as expected. I think it's ready for prod.

**Comment by mattgodbolt** [comment]
Fantastic! Pushig now; will roll out over the next 45m

**Comment by mattgodbolt** [comment]
Thanks again for your patience!

**Comment by mattgodbolt** [comment]
Doops, thanks @RubenRBS

**Pull request merged**
